pkg/azuredisk/nodeserver.go

```diff
@@ -489,3 +510,111 @@ func (d *Driver) ensureBlockTargetFile(target string) error {
+		return nil, status.Errorf(codes.Internal, "Unable to parse capacity %v", err)
+	}
+	if size <= 0 {
+		return nil, status.Errorf(codes.Internal, "size(%v) is illegal, please provide the positive size value", size)
+	}
+
+	parameters := make(map[string]string)
+	parameters["skuname"] = req.GetVolumeContext()["storageAccountType"]
+
+	klog.V(2).Infof("nodePublishEphemeralVolume: creating the managed disk(%s)", ephemeralVolumeId)
+	createVolumeReq := &csi.CreateVolumeRequest{
+		Name: ephemeralVolumeId,
+		CapacityRange: &csi.CapacityRange{
+			RequiredBytes: volumehelper.GiBToBytes(int64(size)),
+		},
+		VolumeCapabilities: []*csi.VolumeCapability{req.GetVolumeCapability()},
+		Parameters:         parameters,
+	}
+	createVolumeResp, err := d.CreateVolume(ctx, createVolumeReq)
+	if err != nil {
+		return nil, err
+	}
+	diskURI := createVolumeResp.Volume.VolumeId
+	defer func() {
+		if err != nil {
+			if err1 := d.cloud.DeleteManagedDisk(diskURI); err1 != nil {
+				klog.Errorf("nodePublishEphemeralVolume: fail to delete volume(%s): %v", diskURI, err1)
+			}
+		}
+	}()
+	klog.V(2).Infof("nodePublishEphemeralVolume: create the managed disk(%s) successfully", ephemeralVolumeId)
+
+	// attach volume to node
+	controllerPublishVolumeReq := &csi.ControllerPublishVolumeRequest{
+		VolumeId:         diskURI,
+		NodeId:           d.NodeID,
+		Readonly:         req.GetReadonly(),
+		VolumeCapability: req.GetVolumeCapability(),
+	}
+
+	klog.V(2).Infof("nodePublishEphemeralVolume: attaching the managed disk(%s) to node(%s)", diskURI, d.NodeID)
+	controllerPublishResp, err := d.ControllerPublishVolume(ctx, controllerPublishVolumeReq)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			controllerUnpublishVolumeReq := &csi.ControllerUnpublishVolumeRequest{
+				VolumeId: diskURI,
+				NodeId:   d.NodeID,
+			}
+			if _, err1 := d.ControllerUnpublishVolume(ctx, controllerUnpublishVolumeReq); err1 != nil {
+				klog.Errorf("nodePublishEphemeralVolume: fail to detach volume(%s): %v", diskURI, err1)
+			}
+		}
+	}()
+	klog.V(2).Infof("nodePublishEphemeralVolume: attach the managed disk(%s) to node(%s) successfully", diskURI, d.NodeID)
+
+	nodeStageVolumeReq := &csi.NodeStageVolumeRequest{
+		VolumeId:          diskURI,
+		PublishContext:    controllerPublishResp.PublishContext,
+		StagingTargetPath: req.GetTargetPath(),
+		VolumeCapability:  req.GetVolumeCapability(),
+		VolumeContext:     req.GetVolumeContext(),
+	}
+	_, err = d.NodeStageVolume(ctx, nodeStageVolumeReq)
+	if err != nil {
+		return nil, err
+	}
+	klog.V(2).Infof("nodePublishEphemeralVolume successfully")
+
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func (d *Driver) nodeUnpublishEphemeralVolume(ctx context.Context, req *csi.NodeUnpublishVolumeRequest) (*csi.NodeUnpublishVolumeResponse, error) {
+	volumeId := req.GetVolumeId()
+	diskName := fmt.Sprintf(ephemeralVolumeIdTemplate, volumeId)
+
+	_, diskURI, err := d.cloud.GetDisk(d.cloud.ResourceGroup, diskName)
```

Comment:
there should be already GetDisk logic in `ControllerUnpublishVolume`, so it's not necessary now.

Reply:
`diskURI ` is still needed in the other part of the function.